lib/internal/modules/esm/loader.js

```diff
@@ -168,100 +211,249 @@ class Loader {
+    );
+
+    job = new ModuleJob(
+      this,
+      url,
+      moduleProvider,
+      parentURL === undefined,
+      inspectBrk
+    );
+
+    this.moduleMap.set(url, job);
+
+    return job;
   }
 
-  hook(hooks) {
-    const {
-      resolve,
-      dynamicInstantiate,
-      getFormat,
-      getSource,
-      transformSource,
-      getGlobalPreloadCode,
-    } = hooks;
+  /**
+   *
+   * @param {string | string[]} specifiers Path(s) to the module
+   * @param {string} [parentURL] Path of the parent importing the module
+   * @returns {object | object[]} A list of module export(s)
+   */
+  async import(specifiers, parentURL) {
+    const wasArr = ArrayIsArray(specifiers);
+    if (!wasArr) specifiers = [specifiers];
 
-    // Use .bind() to avoid giving access to the Loader instance when called.
-    if (resolve !== undefined)
-      this._resolve = FunctionPrototypeBind(resolve, null);
-    if (dynamicInstantiate !== undefined) {
-      process.emitWarning(
-        'The dynamicInstantiate loader hook has been removed.');
-    }
-    if (getFormat !== undefined) {
-      this._getFormat = FunctionPrototypeBind(getFormat, null);
-    }
-    if (getSource !== undefined) {
-      this._getSource = FunctionPrototypeBind(getSource, null);
-    }
-    if (transformSource !== undefined) {
-      this._transformSource = FunctionPrototypeBind(transformSource, null);
-    }
-    if (getGlobalPreloadCode !== undefined) {
-      this._getGlobalPreloadCode =
-        FunctionPrototypeBind(getGlobalPreloadCode, null);
+    const count = specifiers.length;
+    const namespaces = new Array(count);
+
+    for (let i = 0; i < count; i++) {
+      const specifier = specifiers[i];
+      const job = await this.getModuleJob(specifier, parentURL);
+      const { module } = await job.run();
+
+      namespaces[i] = module.getNamespace();
     }
+
+    return wasArr ?
+      namespaces :
+      namespaces[0];
   }
 
-  runGlobalPreloadCode() {
-    if (!this._getGlobalPreloadCode) {
-      return;
+  /**
+   * Provide the ready-to-use source (output is something node understands).
+   *
+   * The internals of this WILL change when chaining is implemented,
+   * depending on the resolution/consensus from #36954
+   * @param {string} url The URL/path of the module to be loaded
+   * @param {Object} context Metadata about the module
+   * @returns {Object}
+   */
+  async load(url, context = {}) {
+    const { 0: defaultLoader, 1: customLoader } = this.#loaders;
+
+    const loader = customLoader ?? defaultLoader;
+    const loaded = await loader(url, context, defaultLoader);
+
+    if (typeof loaded !== 'object') {
+      throw new ERR_INVALID_RETURN_VALUE(
+        'object',
+        'loader load',
+        loaded,
+      );
     }
-    const preloadCode = this._getGlobalPreloadCode();
-    if (preloadCode === null) {
-      return;
+
+    const {
+      format,
+      source,
+    } = loaded;
+
+    if (format === null) {
```

Comment:
Should this be a nullish check instead?
```suggestion
    if (format == null) {
```